### DISCUSSION TITLE
IrReceived generates a lot of unnecessary mqtt messages

### DISCUSSION THREAD

**Discussion by Continuous2Discrete** [arendst/Tasmota#21574] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

I am trying to capture some buttons from my TV remote but only specific ones.

Tasmota immediately publishes any received IR code in MQTT in /tele topic, no matter how TelePeriod is set (I tried 0, 1, 300). A long press on e.g. the volume button sends hundreds of messages in 2-3 seconds, but I don't even need this information, my MQTT broker doesn't need to be notified hundreds of times that I have changed the channel or turned up the TV. 

Instead, I would like to be able to decide which data from IrReceived will be published to MQTT broker using Rules e.g. only if the red button on the TV remote control is pressed publish this information in MQTT and then... HA will close my curtains. HA does not have to check which button was pressed, Tasmota can do that and transmit only what is necessary.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Uniplay Universal Remote IR Controller
- [x] Tasmota binary firmware version number used: 14.1.0 (release-ir)
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: tasmotizer-1.2
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
15:17:57.895 MQT: stat/tasmota_E035B0/RESULT = {"NAME":"Most IR","GPIO":[1,3200,1,3232,576,1088,0,0,320,32,1056,0,0,0],"FLAG":0,"BASE":62}
15:17:58.096 MQT: stat/tasmota_E035B0/RESULT = {"Module":{"0":"Most IR"}}
15:17:58.303 MQT: stat/tasmota_E035B0/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"3200":"Serial Tx"},"GPIO2":{"0":"None"},"GPIO3":{"3232":"Serial Rx"},"GPIO4":{"576":"LedLink_i"},"GPIO5":{"1088":"IRrecv"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"320":"Led_i1"},"GPIO13":{"32":"Button1"},"GPIO14":{"1056":"IRsend1"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
15:27:35.526 MQT: stat/tasmota_E035B0/STATUS = {"Status":{"Module":0,"DeviceName":"Most IR","FriendlyName":["Most IR"],"Topic":"tasmota_E035B0","ButtonTopic":"0","Power":"1","PowerLock":"0","PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
15:27:35.537 MQT: stat/tasmota_E035B0/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota-ir.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:10:34","StartupUTC":"2024-06-05T14:17:01","Sleep":50,"CfgHolder":4617,"BootCount":25,"BCResetTime":"2024-06-04T19:35:27","SaveCount":86,"SaveAddress":"F4000"}}
15:27:35.545 MQT: stat/tasmota_E035B0/STATUS2 = {"StatusFWR":{"Version":"14.1.0(release-ir)","BuildDateTime":"2024-06-03T11:53:03","Boot":31,"Core":"2_7_6","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"391/699"}}
15:27:35.552 MQT: stat/tasmota_E035B0/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["WIFI-2.4GHz",""],"TelePeriod":0,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A00192800000000","00010080","00006000","00404000","00001002"]}}
15:27:35.565 MQT: stat/tasmota_E035B0/STATUS4 = {"StatusMEM":{"ProgramSize":662,"Free":340,"Heap":22,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144068","FlashFrequency":40,"FlashMode":"DOUT","Features":["0809","0F1A0582","04000001","00000003","00000000","00000000","00000000","00000000","04000000","00000080","00000000"],"Drivers":"1,2,!4,5,7,9,10,12,62,!68","Sensors":"1"}}
15:27:35.577 MQT: stat/tasmota_E035B0/STATUS5 = {"StatusNET":{"Hostname":"tasmota-E035B0-5552","IPAddress":"192.168.1.16","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"EC:FA:BC:E0:35:B0","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
15:27:35.585 MQT: stat/tasmota_E035B0/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.1.101","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_E035B0","MqttUser":"Max","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
15:27:35.595 MQT: stat/tasmota_E035B0/STATUS7 = {"StatusTIM":{"UTC":"2024-06-05T14:27:35Z","Local":"2024-06-05T15:27:35","StartDST":"2024-03-31T02:00:00","EndDST":"2024-10-27T03:00:00","Timezone":"+01:00","Sunrise":"04:48","Sunset":"20:48"}}
15:27:35.603 MQT: stat/tasmota_E035B0/STATUS10 = {"StatusSNS":{"Time":"2024-06-05T15:27:35"}}
15:27:35.614 MQT: stat/tasmota_E035B0/STATUS11 = {"StatusSTS":{"Time":"2024-06-05T15:27:35","Uptime":"0T00:10:34","UptimeSec":634,"Vcc":3.489,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"WIFI-2.4GHz","BSSId":"34:6B:46:9E:D4:FF","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:04"}}}

```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
15:36:19.700 CMD: weblog 4
15:36:19.706 MQT: stat/tasmota_E035B0/RESULT = {"WebLog":4}
15:36:20.073 CFG: Saved to flash at F9, Count 89, Bytes 4096
15:36:22.503 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB50AF","DataLSB":"0x20DF0AF5","Repeat":0}}
15:36:23.253 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:25.251 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:26.503 WIF: Checking connection...
15:36:27.501 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB8877","DataLSB":"0x20DF11EE","Repeat":0}}
15:36:37.767 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:39.802 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:40.501 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:41.902 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:45.251 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB50AF","DataLSB":"0x20DF0AF5","Repeat":0}}
15:36:46.301 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB906F","DataLSB":"0x20DF09F6","Repeat":0}}
15:36:46.503 WIF: Checking connection...
15:36:47.201 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x4FB40BF","DataLSB":"0x20DF02FD","Repeat":0}}
15:36:47.298 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:47.448 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:47.713 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:47.848 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:47.948 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.098 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.298 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.398 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.498 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.598 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:36:48.748 MQT: tele/tasmota_E035B0/RESULT = {"IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF","DataLSB":"0xFFFFFFFFFFFFFFFF","Repeat":1}}
15:37:03.240 WIF: Sending Gratuitous ARP
15:37:06.503 WIF: Checking connection...

```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
I would like to be able to decide which data from IrReceived will be published to MQTT broker using Rules.

I found something similar but only for SSerialReceived
> You could activate SetOption147 to disable publishing SSerialReceived MQTT messages. If disabled, you must use event trigger rules instead (SSerialReceived#Data=<string>) to control what, when and how is being published to your MQTT broker or whatever you want.

How can I achieve this with IrReceived?

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
This would add complex logic and code size increase. This is not needed. Every mqtt broker can handle easily this traffic. To be clear this is way easier to handle in the Backend System (HA).

**Comment by Continuous2Discrete** [comment]
Why is it possible to stop MQTT publishing for SSerialReceived (SetOption147), but not for IrReceived? Is that such a big difference? Why does the TelePeriod parameter not work and publication in /tele is always immediate?

**Reply by sfromis** [reply]
TelePeriod is only for telemetry messages, and what happens "now", like IR, is not even related to telemetry, but by necessity has to be immediate. Getting IR codes minutes later would not be really useful.

Your case with a lot of IR messages that way is pretty atypical, but serial will often be part of a flow with constant activity, instead of just when a button is pushed. Pretty big difference.

**Reply by Continuous2Discrete** [reply]
So if it's not even related to telemetry, why is it published in the /tele topic all the time? Is it possible to turn this off? There is an MqttLog where I can decide if I want to publish in /stat topic or not (MqttLog 0/2), so why publish the same thing over and over again without being able to control it in telemetry /tele  topic, which is not even related to it. In the documentation /tele topic is described as: _tele - reports telemetry info at specified intervals._  In my opinion, it doesn't fit together

**Reply by barbudor** [reply]
MqttLog is different. As the name suggest it is for logging, ie debugging. With MqttLog all log message (same as web console or syslog) are pushed to 1 single mqtt topic for recording. This is not for operational usage.

IR Received messages are pushed to MQTT systematically for opérational reason. Tasmota acts as a IR to MQTT gateway, same as BLE to MQTT or Zigbee to MQTT in order an external backend (HA, node-red, io broker....) to act upon those messages.
So pushing those messages is the main goal of the feature

Serial is different in that tasmota is not only used as a basic serial to MQTT gateway but also to interface with sensors or actuators that have a serial interface and for which the serial messages needs to be processed locally with rules or scripting

Of course, IR is also used to control devices such as HVAC with local rule processing but in that case it doesn't receive may messages, only those directed to the device with the remote. So a couple of messages a day doesn't usually another users.

**Reply by sfromis** [reply]
Yes, usage of the `tele` prefix is a bit inconsistent, but what is meant by the telemetry distinction is that it is about the messages published at the telemetry period, at intervals which are arbitrary points in time. The default of every 5 minutes means that the whole cycle could just as well be 2 minutes earlier or later. Telemetry is about monitoring, typically environment sensors of various types, not about things happening here and now for attention.

**Reply by Continuous2Discrete** [reply]
And yet it is possible. I managed to achieve what I expected by compiling a custom version of tasmota. I have extended the operation of SetOption147 so that it also applies to IrReceived. 

It may be primitive but it works. I changed only 3 lines of code in xdrv_05_irremote_full.ino (based on xdrv_08_serial_bridge.ino)
From:
```
ResponseJsonEndEnd();
MqttPublishPrefixTopicRulesProcess_P(RESULT_OR_TELE, PSTR(D_JSON_IRRECEIVED));
```
To:
```
ResponseJsonEndEnd();
if (Settings->flag6.mqtt_disable_sserialrec ) {  // SetOption147  If it is activated, Tasmota will not publish SSerialReceived MQTT messages, but it will proccess event trigger rules
    XdrvRulesProcess(0);
 } else {
    MqttPublishPrefixTopicRulesProcess_P(RESULT_OR_TELE, PSTR(D_JSON_IRRECEIVED));
 }
```

Now I can decide (by SetOption147 0/1) which data from IrReceived will be published to MQTT broker using Rules, what is more optimal for me.

Actually, I never expected a solution to this case to be added as a new feature. I just asked if it was possible and how I could achieve it, but maybe I've been misunderstood. Anyway, thanks for your help.

**Reply by sfromis** [reply]
Sure, with open source, hacking the code is up to your willingness to dive in, instead of just "possible" or not.

Do note that when you open an "issue", that is meant for cases where you see it as a flaw in Tasmota.

**Comment by arendst** [comment]
Extend command ``SetOption147 1`` to disable publish of IRReceived MQTT messages (#21574) implemented in v14.1.0.2
